Add specs on DistributedContext and propagation. (#97)
open-telemetry/opentelemetry-specification: specification/distributedcontext-api.md

An `Entry` is used to label anything that is associated
with a specific operation, such as an HTTP request.

`DistributedContext` is an abstract data type that represents collection of entries.

> i.e., each key is associated with exactly one value. `DistributedContext` is serializable,

Reviewer: ```suggestion
Each key  of `DistributedContext` is associated with exactly one value. `DistributedContext` is serializable,
```

Otherwise sentence starts with the "i.e."

> i.e., each key is associated with exactly one value. `DistributedContext` is serializable,

Reviewer 2: What is meant by key? It's not defined here when reading top to bottom. If it's what I think, then I would write: "`DistributedContext` is an abstract data type that maps `EntryKey`s to `Entry` objects.

Author: Rephrased with Sergey's suggestion.

> and it represents entries that could be propagated inside the process and across process boundaries. 

Reviewer 2: ```suggestion
to facilitate propagating it not only inside the process but also across process boundaries. 
```

> `DistributedContext` is a recommended name but languages can have more language-specific names.

Reviewer: ```suggestion
`DistributedContext` is a recommended name but languages can have more language-specific names like `dctx`.
```

The order of entries is not significant. The iterator MUST not modify the underlying collection of
entries in this `DistributedContext`.

## GetEntryValue

Returns the `EntryValue` associated with the given `EntryKey`, or null if the given `EntryKey`
is not present.

Required parameter:

`EntryKey` entry key to return the value for.

## Limits
Combined size of all entries should not exceed 8192 bytes before encoding.
The size restriction applies to the deserialized entries so that the set of decoded
 `DistributedContext`s is independent of the encoding format.

## DistributedContext Propagation
`DistributedContext` may be propagated across process boundaries or across any arbitrary boundaries for various 
reasons. For example, one may propagate 'project-id' Entry across all micro-services to break down metrics
by 'project-id'. Not all entries in a `DistributedContext` should be propagated and not all entries in a `DistributedContext`
should be accepted from a remote peer. Hence, `DistributedContext` propagator must allow specifying an optional
list of ordered `EntryPropagationFilter`s for receiving entries or for forwarding entries or for both. 

> An `EntryPropagationFilter` list for receiving MAY be different then that for forwarding.

Reviewer 3: typo: `s/then/than`

Author: Good catch, done.

   Current Entrys E1=V1/M1, E2=V2/M2  <== E2 is restored.
Close Scope 1
``` 

# DistributedContext

## GetIterator

Returns an iterator over the entries in this `DistributedContext`.
The order of entries is not significant. The iterator MUST not modify the underlying collection of
entries in this `DistributedContext`.

> ## GetEntryValue

Reviewer 4: Should this be `GetEntry`? `GetEntryValue` seems to return the `EntryValue` which doesn't include the `EntryMetadata`.
```
DistributedContext.GetEntry('name').EntryMetadata
DistributedContext.GetEntry('name').EntryValue
```

Author: The operations are from Java APIs (https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/distributedcontext/DistributedContext.java) and currently there isn't an API `GetEntry`. I think it makes sense to have this `GetEntry` API that looks up entries by name - would like to hear others' opinions.

Reviewer: Please file an issue with suggestion of a better name.

At a minimum, `Exclude` and `Include` actions MUST be implemented.

**Exclude**
If the `EntryPropagationFilterAction` is Exclude then any `Entry` whose `EntryKey` evaluates to true 
with the condition (`EntryPropagationFitlerMatchOperator` and `EntryPropagationFilterMatchString`)
MUST be excluded.

**Include**
If the `EntryPropagationFilterAction` is Include then any `Entry` whose `EntryKey` evaluates to true 
with the condition (`EntryPropagationFitlerMatchOperator  ` and `EntryPropagationFilterMatchString`)
MUST be included. 
  
#### EntryPropagationFilterMatchOperator

| Operator | Description |
|----------|-------------|
| EQUAL | The condition is evaluated to true if `EntryKey` is exactly same as `EntryPropagationFilterMatchString` |
| NOTEQUAL | The condition is evaluated to true if `EntryKey` is NOT exactly same as `EntryPropagationFilterMatchString` |
| HAS_PREFIX | The condition is evaluated to true if `EntryKey` begins with `EntryPropagationFilterMatchString` |

#### EntryPropagationFilterMatchString
It is a string to compare against EntryKey using `EntryPropagationFilterMatchOperator` in order to 

> include or exclude a `Entry`. 

Reviewer 4: nit, a -> an.

   Current Entries E1=V1/M1, E2=V2/M2  <== E2 is restored.
Close Scope 1
``` 

# DistributedContext

## GetIterator

Returns an iterator over the entries in this `DistributedContext`.
The order of entries is not significant. The iterator MUST not modify the underlying collection of
entries in this `DistributedContext`.

## GetEntryValue

Returns the `EntryValue` associated with the given `EntryKey`, or null if the given `EntryKey`
is not present.

Required parameter:

`EntryKey` entry key to return the value for.

## Limits
Combined size of all entries should not exceed 8192 bytes before encoding.
The size restriction applies to the deserialized entries so that the set of decoded
 `DistributedContext`s is independent of the encoding format.

## DistributedContext Propagation

> `DistributedContext` may be propagated across process boundaries or across any arbitrary boundaries for various 

Reviewer 5: Suggestion:
`DistributedContext` may be propagated across any arbitrary boundary (process, $OTHER_BOUNDARY1, $OTHER_BOUNDARY2, etc).

Author: SG, updated.
